### PR TITLE
Refactor: Change equijoin keys from column to expression in logical join

### LINKED ISSUE #4389: Proposal: Improve the join keys of logical plan
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Join keys in `LogicalPlan::Join` are `columns` now, which is the same as physical plan.
To add features base on it, we need add additional projections in logical plan level, like #4193 and #4353. There are some drawbacks:

* Join keys may need alias when do type coercion, and the alias will display in the logical plan (also make display verbose).
* Our logical plan will be verbose with a lot additional projections.
* In optimizing stage, we have to check the additional projections if we want to optimize joins.

### Join key need alias when do type coercion ###
One join key may happen one more time in join on condition, then it may also need do type coercion one more time, to distinguish each position, we have to add alias for them.
For example, we have:
```sql
--- t0.a t1.a -> UInt8
--- t0.b t1.b -> Int16
--- t0.c t1.c -> Int32
select t0.a, t1.a from test0 as t0
                  inner join test1 as t1
                  on t0.a = t1.b and t0.a = t1.c      
```

then the logical plan will:
```
-- before optimizer
Projection: t0.a, t1.a                                                                                                                                                                                 
 Inner Join: t0.a = t1.b, t0.a = t1.c                                                                                                                                                             
  Projection: t0.a, t0.b, t0.c                                                                                                                   
   SubqueryAlias: t0                                                                                                                                                                                
    TableScan: test0 projection=[a,b,c]                                                                                                                                                               
   SubqueryAlias: t1 

-- after type coercion
Projection: t0.a, t1.a                                                                                                                                                                                 
 Inner Join: t0.a#0 = t1.b, t0.a#1 = t1.c                                                                                                                                                             
  Projection: t0.a, CAST(t0.a AS Int16) AS t0.a#0, CAST(t0.a AS Int32) AS t0.a#1                                                                                                                     
   SubqueryAlias: t0                                                                                                                                                                                
    TableScan: test0 projection=[a]                                                                                                                                                               
   SubqueryAlias: t1                                                                                                                                                                                 
```

This makes our logical plan more complex.

### logical plan is verbose ###
for sql:
```sql
select t0.a, t1.b
       from test0 as t0
       inner join test1 as t1
       on t0.a + 1 = t1.a * 2;
```
the logical plan will be like(does not do type coercion):
```
Projection: t0.a, t1.b                                                                                                                          
 Inner Join: t0.a + Int64(1) = t1.a * Int64(2)                                                                                                        
  Projection: t0.a, CAST(t0.a AS Int64) + Int64(1)                                                                                                  
   SubqueryAlias: t0                                                                                                                                
    TableScan: test0 projection=[a]                                                                                                               
  Projection: t1.b, CAST(t1.a AS Int64) * Int64(2)                                                                                                   
   SubqueryAlias: t1                                                                                                                                
    TableScan: test1 projection=[a, b]                                                                                                             
```

Two projections will be added before join.

**Describe the solution you'd like**
I would like to fold the additional projections to join(logical-plan), and keep the physical join as before. I think it can make our logical plan clean and easy to extend.
```rust
/// Join two logical plans on one or more join columns
#[derive(Clone)]
pub struct Join {
    ...
    /// Equijoin clause expressed as pairs of (left, right) join columns
    pub on: Vec<(Expr,Expr)>,
   ...
}
``` 

After I check the current code base, the main changes are the optimizers:

* eliminate_cross_join.
* filter_null_join_key.
* filter_push_down.
* projection_pushdown.
* subquery_filter_to_join.

We can step by step to finish it, and refactor the LogicalPlan::Join finally.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Closes #4389.

# Rationale for this change

It can make the display representation of logical join more clean.

# What changes are included in this PR?

* Change equijoin keys from column to expression in logical join.
* Remove additional projection in `parse_join`.
* Add additional physical projection for physical join.
* Adapter the expression keys in `EliminateCrossJoin`, `FilterNullJoinKeys`, `PushDownFilter`, `SubqueryFilterToJoin` optimization rules.
* Change the equijoin keys of `JoinNode` in `proto` module.
* Fix and add test cases. 

# Are these changes tested?
Yes

# Are there any user-facing changes?
Yes, the `JoinNode` in `proto` module is also changed.

### CODE REVIEW

File: datafusion/core/tests/sql/joins.rs [hunk 1]
```diff
@@ -2448,3 +2492,289 @@ async fn reduce_cross_join_with_wildcard_and_expr() -> Result<()> {
...
+        let msg = format!("Creating logical plan for '{}'", sql);
+        let plan = ctx.create_logical_plan(sql).expect(&msg);
+        let state = ctx.state();
+        let logical_plan = state.optimize(&plan)?;
+        let physical_plan = state.create_physical_plan(&logical_plan).await?;
+
+        let expected = if repartition_joins {
+            vec![
+                "ProjectionExec: expr=[t1_id@0 as t1_id, t2_id@2 as t2_id, t1_name@1 as t1_name]",
+                "  ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@1 as t1_name, t2_id@3 as t2_id]",
```
Comment (ygf11) [inline datafusion/core/tests/sql/joins.rs:2516]: It seems merging projection does not consider the ordering also.

File: datafusion/core/tests/sql/joins.rs [hunk 2]
```diff
@@ -2432,6 +2428,54 @@ async fn reduce_cross_join_with_wildcard_and_expr() -> Result<()> {
...
+        let expected = if repartition_joins {
+            vec![
+                "ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@1 as t1_name, t1_int@2 as t1_int, t2_id@3 as t2_id, t2_name@4 as t2_name, t2_int@5 as t2_int, CAST(t1_id@0 AS Int64) + 11 as t1.t1_id + Int64(11)]",
+                "  ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@1 as t1_name, t1_int@2 as t1_int, t2_id@4 as t2_id, t2_name@5 as t2_name, t2_int@6 as t2_int]",
+                "    CoalesceBatchesExec: target_batch_size=4096",
+                "      HashJoinExec: mode=Partitioned, join_type=Inner, on=[(Column { name: \"t1.t1_id + Int64(11)\", index: 3 }, Column { name: \"CAST(t2.t2_id AS Int64)\", index: 3 })]",
+                "        CoalesceBatchesExec: target_batch_size=4096",
+                "          RepartitionExec: partitioning=Hash([Column { name: \"t1.t1_id + Int64(11)\", index: 3 }], 2)",
+                "            ProjectionExec: expr=[t1_id@0 as t1_id, t1_name@1 as t1_name, t1_int@2 as t1_int, CAST(t1_id@0 AS Int64) + 11 as t1.t1_id + Int64(11)]",
+                "              RepartitionExec: partitioning=RoundRobinBatch(2)",
```
Comment (ygf11) [inline datafusion/core/tests/sql/joins.rs:2447]: The name in join and projection matches now, both are `t1.t1_id + Int64(11)`.
cc @liukun4515

**Comment by ygf11** [comment]
@alamb @mingmwang @jackwener @liukun4515, please take a look, thanks.

**Comment by alamb** [comment]
I plan to review this carefully tomorrow.

File: datafusion/proto/src/logical_plan.rs [hunk 3]
```diff
@@ -727,17 +733,24 @@ impl AsLogicalPlan for LogicalPlanNode {
...
                         (left_keys, right_keys),
                         filter,
                     )?,
-                    JoinConstraint::Using => builder.join_using(
-                        &into_logical_plan!(join.right, ctx, extension_codec)?,
-                        join_type.into(),
-                        left_keys,
-                    )?,
+                    JoinConstraint::Using => {
+                        // The equijoin keys in using-join must be column.
```
Comment (alamb) [inline datafusion/proto/src/logical_plan.rs:743]: 👍

File: datafusion/sql/src/planner.rs [hunk 4]
```diff
@@ -6074,12 +6055,10 @@ mod tests {
             ON orders.customer_id * 2 = person.id + 10";
 
         let expected = "Projection: person.id, orders.order_id\
-        \n  Projection: person.id, person.first_name, person.last_name, person.age, person.state, person.salary, person.birth_date, person.😀, orders.order_id, orders.customer_id, orders.o_item_id, orders.qty, orders.price, orders.delivered\
-        \n    Inner Join: person.id + Int64(10) = orders.customer_id * Int64(2)\
-        \n      Projection: person.id, person.first_name, person.last_name, person.age, person.state, person.salary, person.birth_date, person.😀, person.id + Int64(10)\
-        \n        TableScan: person\
-        \n      Projection: orders.order_id, orders.customer_id, orders.o_item_id, orders.qty, orders.price, orders.delivered, orders.customer_id * Int64(2)\
-        \n        TableScan: orders";
+        \n  Inner Join: person.id + Int64(10) = orders.customer_id * Int64(2)\
```
Comment (alamb) [inline datafusion/sql/src/planner.rs:6058]: this plan is certainly much nicer -- I don't understand where all the other columns used to come from but this is 👍

File: datafusion/core/tests/sql/joins.rs (hunk 1 above)
Comment (alamb) [reply datafusion/core/tests/sql/joins.rs:2516]: 👍 I think that could be improved as a follow on PR

**Review by alamb — APPROVED** [review]
This is a really cool PR. Thank you @ygf11 

I went through the code and all the test changes carefully and I think this PR is 👨‍🍳 👌  very nice

I think it is ready to go, though plan to leave it open for another day or two in case @jackwener  or anyone else would like a chance to review it.

**Comment by alamb** [comment]
Because this PR has been outstanding for some time, and it had a conflict I took the liberty to merge from master and resolve the conflicts

I am about out of time today but I plan to merge this PR first thing tomorrow. 

Thanks again @ygf11  -- really great stuff.

File: datafusion/core/tests/sql/joins.rs (hunk 1 above)
Comment (ygf11) [reply datafusion/core/tests/sql/joins.rs:2516]: I will work on it.

**Comment by ygf11** [comment]
Thanks @alamb, I resolve the remaining conflict, and ci success now.

**Comment by alamb** [comment]
Thanks again @ygf11  -- this is great work

**Pull request merged**